### ISSUE TITLE
Regression: After #128 got fixed all my messages are sent twice when they've been sent from "real" WhatsApp app

### ISSUE THREAD

**Issue by wapsi** [tulir/mautrix-whatsapp#132]
Regression: After #128 got fixed all my messages are sent twice when they've been sent from "real" WhatsApp app. The duplicated messages are not shown in Matrix/Riot but I'm seeing them on my phone's WhatsApp app and all other members in my WA groups are seeing all my messages sent twice.

A workaraound is not to use Matrix puppet takeover (by issuing a command: logout-matrix).

**Comment by tulir** [comment]
The problem was some event ID conflict. Recreating the portals seemed to fix it and I couldn't reproduce it afterwards.

**Comment by wapsi** [comment]
Hmm, OK. How can I recreate all portals easily? By using logout & login (WhatsApp) using the bot?

**Comment by wapsi** [comment]
I attached debug level log files from maturix-whatsapp and matrix-synapse. It's a case where I sent message "test6" from my phone's WhatsApp app and right after mautrix-whatsapp / matrix-synapse was handled it => the duplicated message appeared on my phone:
```
test6
test6
```

Maybe the interesting part from the logs is:
```
Jan 05 11:22:42 matrix mautrix-whatsapp[22502]: [Jan  5, 2020 11:22:42] [Database/Message/WARN] Failed to insert 📞-📞@g.us@<message-id 2 was here>: pq: duplicate key value violates unique constraint "message_mxid_key"
Jan 05 11:22:42 matrix mautrix-whatsapp[22502]: [Jan  5, 2020 11:22:42] [Portal/📞-📞@g.us/DEBUG] Sending event <another message-id was here> to WhatsApp
Jan 05 11:22:43 matrix mautrix-whatsapp[22502]: [Jan  5, 2020 11:22:43] [User/@MyMatrixNickname:my.matrix.fi/DEBUG] JSON message: ["MsgInfo",{"cmd":"ack","id":"<message-id was here>","ack":2,"from":"📞@c.us","to":"📞-📞@g.us","t":📞,"participant":"📞@c.us"}]
Jan 05 11:22:43 matrix mautrix-whatsapp[22502]: [Jan  5, 2020 11:22:43] [Portal/📞-📞@g.us/DEBUG] Handled Matrix event: &{<nil> @MyMatrixNickname:my.matrix.fi {m.room.message 0} 📞 <another message-id was here> <matrix room-id was here>:my.matrix.fi {[<value cut>] map[body:test6 msgtype:m.text] m.text test6   <nil>   <nil> <nil> <nil> {   false <nil> } []    map[] []}  {<nil>   96  {map[] {{[] false 0} map[]} {{[] false 0} []} {{[] false 0} []}}  <nil>} []}
Jan 05 11:22:43 matrix mautrix-whatsapp[22502]: [Jan  5, 2020 11:22:43] [User/@MyMatrixNickname:my.matrix.fi/DEBUG] JSON message: ["MsgInfo",{"cmd":"ack","id":"<message-id 2 was here>","ack":2,"from":"📞@c.us","to":"📞-📞@g.us","t":📞,"participant":"📞@c.us"}]
```

[duplicatemessagebug.log](https://github.com/tulir/mautrix-whatsapp/files/4023083/duplicatemessagebug.log)

**Comment by wapsi** [comment]
I think I found the reason for this issue: The mautrix-whatsapp believes the message which was just sent by itself to Matrix is a new message (because Matrix is notifying mautirx-whatsapp about the message it just sent there) and it processes it and sends it to WhatsApp as new message. And because this same message already exists in Matrix and in mautrix-whatsapp DBs it throws this warning:
`[Database/Message/WARN] Failed to insert 📞-📞@g.us@<message-id 2 was here>: pq: duplicate key value violates unique constraint "message_mxid_key"`
But mautrix-whatsapp doesn't care about this and it sends it to WhatsApp anyway and that's why the message is duplicated in WhatsApp app and not in the Matrix.

I created (kind of) a workaround in portal.go:

New function:
```
func (portal *Portal) checkIfMxidExists(mxid types.MatrixEventID) bool {
        msg := portal.bridge.DB.Message.GetByMXID(mxid)
        if msg != nil {
                return true
        }
        return false
}
```
And then in
func (portal *Portal) HandleMatrixMessage(sender *User, evt *mautrix.Event) function at the very beginning I call it:
```
        if (portal.checkIfMxidExists(evt.ID)) {
                portal.log.Debugln("Received event with mxid:", evt.ID, "but it has been processed earlier (most probably loopback message), skipping it")
                return
        }
```

With these modifications I'm not sending messages twice anymore but I think it's not very good fix because it doesn't fix the root cause of this bug. The mautrix-whatsapp should not handle the messages which have been sent by itself to Matrix as new messages (and "re"-send them back to WhatsApp).

Maybe @tulir knows better why this happens and how this could be fixed correctly.
